ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/TransactionLogUpdatesConversions.scala

```diff
@@ -0,0 +1,423 @@
+                    transactionId = first.transactionId,
+                    commandId = getCommandId(
+                      events,
+                      filter.keySet,
+                    ),
+                    workflowId = first.workflowId,
+                    effectiveAt = Some(instantToTimestamp(first.ledgerEffectiveTime)),
+                    events = flatEvents,
+                    offset = ApiOffset.toApiString(tx.offset),
+                    traceContext = None,
+                  )
+                )
+              } else None
+            }
+        }
+        .getOrElse(Future.successful(None))
+    }
+
+    private def removeTransient(aux: Vector[TransactionLogUpdate.Event]) = {
+      val permanent = aux.foldLeft(Set.empty[ContractId]) {
+        case (contractIds, event) if !contractIds(event.contractId) =>
+          contractIds + event.contractId
+        case (contractIds, event: ExercisedEvent) if event.consuming =>
+          contractIds - event.contractId
+        case (contractId, _) =>
+          throw new RuntimeException(s"Unexpected non-consuming event for contractId $contractId")
+      }
+      aux.filter(ev => permanent(ev.contractId))
+    }
+
+    private val FlatTransactionPredicate =
+      (event: TransactionLogUpdate.Event, filter: FilterRelation) => {
+        val (parties, partiesTemplates) = filter.partition(_._2.isEmpty)
+        val wildcardParties = parties.keySet.map(_.toString)
+
+        val templateSpecificParties = partiesTemplates
+          .foldLeft(Map.empty[Ref.Identifier, mutable.Builder[String, Set[String]]]) {
+            case (acc, (k, vs)) =>
+              vs.foldLeft(acc) { case (a, v) =>
+                a + (v -> (a.getOrElse(v, Set.newBuilder) += k))
+              }
+          }
+          .view
+          .map { case (k, v) => k -> v.result() }
+          .toMap
+
+        val witnessesMatchingWildcards =
+          event.flatEventWitnesses.intersect(wildcardParties).nonEmpty
+        lazy val templateSpecificPartiesMatchingTemplateId = templateSpecificParties
+          .get(event.templateId)
+          .exists(_.intersect(event.flatEventWitnesses).nonEmpty)
+
+        witnessesMatchingWildcards || templateSpecificPartiesMatchingTemplateId
+      }
+
+    private def toFlatEvent(
+        event: TransactionLogUpdate.Event,
+        requestingParties: Set[Party],
+        verbose: Boolean,
+        lfValueTranslation: LfValueTranslation,
+    )(implicit
+        loggingContext: LoggingContext,
+        executionContext: ExecutionContext,
+    ): Future[Option[apiEvent.Event]] =
+      event match {
+        case createdEvent: TransactionLogUpdate.CreatedEvent =>
+          createdToFlatEvent(requestingParties, verbose, lfValueTranslation, createdEvent)
+
+        case exercisedEvent: TransactionLogUpdate.ExercisedEvent if exercisedEvent.consuming =>
+          Future.successful(Some(exercisedToFlatEvent(requestingParties, exercisedEvent)))
```

Comment:
Please unify the signatures for `exercisedToFlatEvent` and `createdToFlatEvent` so that both return a Future or none returns a Future

Reply:
Are you sure? I think it is better to let the types speak for themselves than to align them:
* `createdToFlatEvent` performs some time-consuming computation that needs an async treatment
* `exercisedToFlatEvent` performs a simple DTO mapping, sync. 

If I were to align them, I would have to force `exercisedToFlatEvent` to return `Future`. However, imho, the new method prototype would be misleading, since it would be hinting some async stuff will be happening under the hood.